codegen/src/main/java/software/amazon/awssdk/codegen/poet/rules/EndpointParamsKnowledgeIndex.java

```diff
@@ -0,0 +1,190 @@
+        TypeName typeName = TypeName.get(AccountIdEndpointMode.class);
+
+        MethodSpec.Builder builder = MethodSpec.methodBuilder("resolveAccountIdEndpointMode")
+                                               .addModifiers(PRIVATE)
+                                               .addParameter(SdkClientConfiguration.class, "config")
+                                               .returns(typeName);
+
+        builder.addStatement("$T configuredMode = config.option($T.$L)", typeName, AwsClientOption.class, keyName);
+
+        builder.beginControlFlow("if (configuredMode == null)");
+        builder.addCode("configuredMode = $T.create()", AccountIdEndpointModeResolver.class);
+        builder.addCode(".profileFile(config.option($T.PROFILE_FILE_SUPPLIER))", SdkClientOption.class);
+        builder.addCode(".profileName(config.option($T.PROFILE_NAME))", SdkClientOption.class);
+        builder.addCode(".defaultMode($T.PREFERRED)", typeName);
+        builder.addStatement(".resolve()");
+        builder.endControlFlow();
+
+        builder.addStatement("return configuredMode");
+        return Optional.of(builder.build());
+    }
+
+    private Map<BuiltInParameter, LocalParameter> builtInsForClientBuilder(Map<String, ParameterModel> serviceEndpointParams) {
+        Map<BuiltInParameter, LocalParameter> actualParams = new EnumMap<>(BuiltInParameter.class);
+        serviceEndpointParams.forEach((k, v) -> {
+            BuiltInParameter builtInEnum = v.getBuiltInEnum();
+            if (builtInEnum != null && BUILT_IN_PARAMS_FOR_CLIENT_BUILDER.containsKey(builtInEnum)) {
+                actualParams.put(builtInEnum, BUILT_IN_PARAMS_FOR_CLIENT_BUILDER.get(builtInEnum));
+            }
+        });
+        return actualParams;
+    }
+
+    public Optional<MethodSpec> accountIdFromIdentityMethod() {
+        if (!hasAccountIdEndpointModeBuiltIn()) {
+            return Optional.empty();
+        }
+
+        ParameterizedTypeName paramType = ParameterizedTypeName.get(ClassName.get(SelectedAuthScheme.class),
+                                                                    TypeVariableName.get("T"));
+
+        MethodSpec.Builder builder = MethodSpec.methodBuilder("accountIdFromIdentity")
+                                               .addModifiers(PRIVATE, STATIC)
+                                               .addTypeVariable(TypeVariableName.get("T", Identity.class))
+                                               .addParameter(paramType, "selectedAuthScheme")
+                                               .returns(String.class);
+
+        builder.addStatement("$T identity = $T.joinLikeSync(selectedAuthScheme.identity())", TypeVariableName.get("T"),
+                             CompletableFutureUtils.class);
+        builder.addStatement("$T accountId = null", String.class);
+        builder.beginControlFlow("if (identity instanceof $T)", AwsCredentialsIdentity.class);
```

Comment:
Just curious, will `AwsCredentialsIdentity` be the only identity type that supports account ID?

Reply:
We'd have to look in the crystal ball for that one. Nothing in the pipelines as for now, but it would be good to have another round of discussions here. It's a dilemma to build at the right level - neither adding support that's super generic and never diversified nor being too specific and then getting a broader use case is great.

Reply:
If we have to add accountId in the future, we can support it through an interface.